src/imitation/algorithms/bc.py

```diff
@@ -4,181 +4,176 @@
+
+from imitation.algorithms.base import (
+    AnyTransitions,
+    DemonstrationAlgorithm,
+    TransitionMapping,
+    make_data_loader,
+)
 from imitation.data import rollout, types
-from imitation.policies import base as policy_base
+from imitation.policies.base import FeedForward32Policy
 from imitation.util import logger
 
 
-def reconstruct_policy(
-    policy_path: str,
-    device: Union[th.device, str] = "auto",
-) -> policies.ActorCriticPolicy:
-    """Reconstruct a saved policy.
+@dataclass(frozen=True)
+class BehaviorCloningLoss:
+    """Container for the different aspects of behavior cloning loss."""
 
-    Args:
-        policy_path: path where `.save_policy()` has been run.
-        device: device on which to load the policy.
+    neglogp: th.Tensor
+    entropy: th.Tensor
+    ent_loss: th.Tensor
+    prob_true_act: th.Tensor
+    l2_norm: th.Tensor
+    l2_loss: th.Tensor
+    loss: th.Tensor
 
-    Returns:
-        policy: policy with reloaded weights.
-    """
-    policy = th.load(policy_path, map_location=utils.get_device(device))
-    assert isinstance(policy, policies.ActorCriticPolicy)
-    return policy
 
+@dataclass(frozen=True)
+class BehaviorCloningLossCalculator:
+    """Functor to compute the loss used in Behavior Cloning."""
 
-class ConstantLRSchedule:
-    """A callable that returns a constant learning rate."""
+    ent_weight: float
+    l2_weight: float
 
-    def __init__(self, lr: float = 1e-3):
-        """Builds ConstantLRSchedule.
+    def __call__(
+        self,
+        policy: ActorCriticPolicy,
+        obs: Union[th.Tensor, np.ndarray],
+        acts: Union[th.Tensor, np.ndarray],
+    ) -> BehaviorCloningLoss:
+        """Calculate the supervised learning loss used to train the behavioral clone.
 
         Args:
-            lr: the constant learning rate that calls to this object will return.
+            policy: The actor-critic policy of which to compute the loss.
+            obs: The observations seen by the expert.
+            acts: The actions taken by the expert.
+
+        Returns:
+            loss: The supervised learning loss for the behavioral clone to optimize.
+            stats_dict: Statistics about the learning process to be logged.
+
         """
-        self.lr = lr
+        _, log_prob, entropy = policy.evaluate_actions(obs, acts)
+        prob_true_act = th.exp(log_prob).mean()
+        log_prob = log_prob.mean()
+        entropy = entropy.mean()
+
+        l2_norms = [th.sum(th.square(w)) for w in policy.parameters()]
+        l2_norm = sum(l2_norms) / 2  # divide by 2 to cancel with gradient of square
 
-    def __call__(self, _):
-        """Returns the constant learning rate."""
-        return self.lr
+        ent_loss = -self.ent_weight * entropy
+        neglogp = -log_prob
+        l2_loss = self.l2_weight * l2_norm
+        loss = neglogp + ent_loss + l2_loss
 
+        return BehaviorCloningLoss(
+            neglogp,
```

Comment:
Maybe use kwargs? Would be easy to mix up positional order.

Reply:
Thx fixed.